oscore-cf/src/main/java/org/eclipse/californium/oscore/OSSerializer.java

```diff
@@ -0,0 +1,557 @@
+			int optionDeltaNibble = getOptionNibble(optionDelta);
+			writer.write(optionDeltaNibble, CoAP.MessageFormat.OPTION_DELTA_BITS);
+
+			// write 4-bit option length
+			int optionLength = option.getLength();
+			int optionLengthNibble = getOptionNibble(optionLength);
+			writer.write(optionLengthNibble, CoAP.MessageFormat.OPTION_LENGTH_BITS);
+
+			// write extended option delta field (0 - 2 bytes)
+			if (optionDeltaNibble == 13) {
+				writer.write(optionDelta - 13, 8);
+			} else if (optionDeltaNibble == 14) {
+				writer.write(optionDelta - 269, 16);
+			}
+
+			// write extended option length field (0 - 2 bytes)
+			if (optionLengthNibble == 13) {
+				writer.write(optionLength - 13, 8);
+			} else if (optionLengthNibble == 14) {
+				writer.write(optionLength - 269, 16);
+			}
+
+			// write option value
+			writer.writeBytes(option.getValue());
+
+			// update last option number
+			previousOptionNumber = option.getNumber();
+		}
+	}
+
+	/**
+	 * Read the E options from an OSCore plaintext (decrypted).
+	 * 
+	 * @param plaintext the OSCore plaintext
+	 * 
+	 * @return the list of E options
+	 */
+	public static List<Option> readConfidentialOptions(byte[] plaintext) {
+		if (plaintext != null) {
+			DatagramReader reader = new DatagramReader(plaintext);
+			List<Option> optionList = new ArrayList<>();
+			int currentOption = 0;
+			byte nextByte = 0;
+			while (reader.bytesAvailable()) {
+				nextByte = reader.readNextByte();
+				if (nextByte != CoAP.MessageFormat.PAYLOAD_MARKER) {
+					// the first 4 bits of the byte represent the option delta
+					int optionDeltaNibble = (0xF0 & nextByte) >> 4;
+					currentOption += readOptionValueFromNibble(reader, optionDeltaNibble);
+
+					// the second 4 bits represent the option length
+					int optionLengthNibble = (0x0F & nextByte);
+					int optionLength = readOptionValueFromNibble(reader, optionLengthNibble);
+
+					// read option
+					Option option = new Option(currentOption);
+					option.setValue(reader.readBytes(optionLength));
+
+					// add option to message
+					optionList.add(option);
+				} else
+					break;
+			}
+			return optionList;
+		} else {
+			LOGGER.severe(OSExceptionHandler.BYTE_ARRAY_NULL);
+			throw new IllegalArgumentException(OSExceptionHandler.BYTE_ARRAY_NULL);
+		}
+	}
+
+	/**
+	 * Read the payload from an OSCore plaintext (decrypted), if any.
+	 * 
+	 * @param plaintext the OSCore plaintext
+	 * 
+	 * @return the payload or null if there is none
+	 * @throws OSException if the payload marker is present and the payload is
+	 *             zero length
+	 */
+	public static byte[] readPayload(byte[] plaintext) throws OSException {
+		if (plaintext != null) {
+			DatagramReader reader = new DatagramReader(plaintext);
+			byte nextByte = 0;
+			while (reader.bytesAvailable()) {
+				nextByte = reader.readNextByte();
+				if (nextByte == CoAP.MessageFormat.PAYLOAD_MARKER) {
+					// the presence of a marker followed by a zero-length
```

Comment:
That's not working!
The value of an option can also contain a CoAP.MessageFormat.PAYLOAD_MARKER (0xff)!

Reply:
Are you sure? This is exactly how the OSCORE draft details how to diff between the E options and the payload. See subchapter 5.3 in https://core-wg.github.io/oscoap/draft-ietf-core-object-security.html#rfc.section.5.3

Reply:
Only the implementation is broken, not the spec.!
When parsing the options, the 0xff at the start of the next option indicates the payload! 0xff within the options may belong to the option value. Consider a ETAG 0x0102FF, this would encode it's value with { 0x0x, 0x02, 0xff}. The parser from californium will read that value and then check for the bytes behind the 0xff, if there is the next option. Your implementation starts at the begin and stops at the first 0xff, so accidental within the option value for the ETAG. I guess, add a unit test with the above ETAG would show the defect. I think, adopting the californium `DataParser` should help to fix that. `parseOptionsAndPayload` checks only for 0xff at the start of the next option.

Reply:
With that ETAG in the server example:

```
			public void handleGET(CoapExchange exchange) {
				System.out.println("Accessing hello/1 resource");
				Response r = new Response(ResponseCode.CONTENT);
				r.setPayload("Hello World!");
				r.getOptions().addETag(new byte[] {0x01, (byte)0xff, 'b', 'u', 'g'});
				exchange.respond(r);
				server.destroy();
			}

```

The client gets:

> RESPONSE CODE: CONTENT 2.05
> RESPONSE PAYLOAD: 62 75 67 ff 48 65 6c 6c 6f 20 57 6f 72 6c 64 21 
> RESPONSE TEXT: bug�Hello World!

Reply:
Thanks for elaborating on the issue. Yes, I like your idea of using the californium DataParser. I will use that instead.

Reply:
With PR #660 the `readPayload` and `readConfidentialOptions` are obsolete.
See comments in specialized decryptor implementations.

Reply:
Great!